### ISSUE TITLE
[feat] Sort by recent

### ISSUE THREAD

**Issue by Gazareth** [GnikDroy/projections.nvim#18]
Hi and thank you for the amazing plugin!

Just wondering if it is possible to sort the projects in the Telescope window by most recently used? Or any other kind of sorting I could leverage?

Thanks again

**Comment by GnikDroy** [comment]
I think telescope has sorters, and it is to be setup from there. I will add a wiki page to do this and respond later today.

**Comment by GnikDroy** [comment]
# Override options (including sorter)
```lua
require('telescope').load_extension('projections')
vim.keymap.set("n", "<leader>fp", function()
    local find_projects = require("telescope").extensions.projections.projections
    find_projects({
        layout_config = {
            width = { 0.80, max = 80 },
            height = { 0.80, max = 20 }
        }, -- This is how you can change the layout
        prompt_title = "Hello There", -- Changing the prompt here
        sorter = -- ****** Write your custom sorter here!! **********
        -- you can override other options as well. check out telescope.nvim for more options.
    })
end, { desc = "Keybinding to find project" })
```

Take a look at sorters provided by telescope for additional information. You can also write your own.

# How to implement "sort by recent"

Firstly, recognize that this only applies at the very beginning. Because as soon as you write something, a fuzzy sort algorithm must be applied. You can check the file modification time of session files present in `stdpath('data') .. 'projection_sessions/'` to sort by recently used. 

One example of how you can find the most recent project is already present in https://github.com/GnikDroy/projections.nvim/blob/fe67b497aad7253a06cd0c07605fe391ea6a2e66/lua/projections/session.lua#L86

# Can this be shipped with projections?

Perhaps. Although this has questionable utility (since you are expected to fuzzy find your project and not navigate the project list manually), this never hurts the user experience. So, I'm willing to make this the default.

**Comment by Gazareth** [comment]
Thanks, I appreciate this. However, after looking into it a bit more, it seems creating a custom sorter is not that easy.

And it might be overkill too. I think we probably want something more akin to [this PR in `telescope-project.nvim`](https://github.com/nvim-telescope/telescope-project.nvim/pull/112).

They are passing the projects to the picker in a pre-sorted table (assuming the config var is unset, or set to 'recent').

What do you think?

**Comment by GnikDroy** [comment]
@Gazareth Sure, as I've said before, I'm willing to make this the default. I won't be adding a config variable, though.

A sorter is the correct way to do this, like changing the prompt and other things. It should allow, among other things, to change sorting behavior at runtime with a keybinding in the telescope window.

I will attempt this later today.

**Comment by GnikDroy** [comment]
I will make this default, so no setup is required. But, setting up a sorter is not very difficult.

```lua
-- Bind <leader>pp to Telescope projections (with recent sort)
vim.keymap.set("n", "<leader>pp", function()
    local find_projects = require("telescope").extensions.projections.projections
    
    -- Sort by recent implemented by
    -- hooking into the scoring function, of generic_sorter
    -- we check if prompt is empty, if so, score by file modification time (inverted)
    -- if prompt is not empty, use the default scoring function
    local config = require("projections.config").config
    local Session = require("projections.session")
    local sorter = require("telescope.config").values.generic_sorter()
    local default_scoring_function = sorter.scoring_function
    sorter.scoring_function = function(_sorter, prompt, line, ...)
        if prompt == '' then
            local session_filename = Session.session_filename(vim.fn.fnamemodify(line, ":h"),
                vim.fs.basename(line))
            return 1 / math.abs(vim.fn.getftime(tostring(config.sessions_directory .. session_filename)))
        end
        return default_scoring_function(_sorter, prompt, line, ...)
    end
    find_projects({ sorter = sorter })
end, { desc = "Find project" })
```

`Session.session_filename` is not exposed yet but is a simple function. This will be exposed when I implement this.
```lua
function Session.session_filename(workspace_path, project_name)
    local path_hash = utils._fnv1a(tostring(workspace_path))
    return string.format("%s_%u.vim", project_name, path_hash)
end
```

**Comment by GnikDroy** [comment]
Sort by recent is now the default on empty prompts. Fixed by 67f80a2a44bb425849f966db57a53415d478c5ae

Feel free to reopen if you have any issues
